src/datachain/lib/signal_schema.py

```diff
@@ -749,3 +749,120 @@ def _build_tree_for_model(
+                            f"Column {column} not found in the schema"
+                        )
+
+                    # Keep the first part of the column name (first signal name)
+                    # as it will be used to store the partial type in the schema.
+                    parent_type_origin = serialized[signal]
+                    parent_type_partial = type_name_to_partial(parent_type_origin)
+
+                    # Store the first signal name in the schema
+                    schema[signal] = parent_type_partial
+                    continue
+
+                # Process the rest of the column name (signal names),
+                # which should be in the custom types.
+                if parent_type_origin not in custom_types:
+                    raise SignalSchemaError(
+                        f"Custom type {parent_type_origin} not found in the schema"
+                    )
+
+                # Get the custom type for the parent signal and check if the signal
+                # is in the custom type fields.
+                custom_type = custom_types[parent_type_origin]
+                signal_type = custom_type["fields"].get(signal)
+                if not signal_type:
+                    raise SignalSchemaError(
+                        f"Field {signal} not found in custom type {parent_type_origin}"
+                    )
+
+                partial_type = type_name_to_partial(signal_type)
+
+                if parent_type_partial in schema["_custom_types"]:
+                    # If the parent type is already in the schema, just add the signal
+                    # to the custom types.
+                    schema["_custom_types"][parent_type_partial]["fields"][signal] = (
+                        partial_type
+                    )
+                else:
+                    # If the parent type is not in the schema, add it
+                    # to the custom types with the signal.
+                    schema["_custom_types"][parent_type_partial] = {
+                        "schema_version": 2,
+                        "name": parent_type_partial,
+                        "fields": {signal: partial_type},
+                        "bases": [
+                            (get_type_name(partial_type), "__main__", partial_type),
+                            ("DataModel", "datachain.lib.data_model", "DataModel@v1"),
+                            ("BaseModel", "pydantic.main", None),
+                            ("object", "builtins", None),
+                        ],
+                    }
+
+                parent_type_origin, parent_type_partial = signal_type, partial_type
+
+        if not schema["_custom_types"]:
+            del schema["_custom_types"]
+
+        return SignalSchema.deserialize(schema)
+
+
+def type_name_to_partial(type_: str) -> str:
+    """Convert a type name to a partial type name with the same version."""
+    if "@" in type_:
+        type_name, version = type_.split("@")
```

Comment:
can in some cases version be empty (just looking into `parse_name_version`)

Reply:
I can not think of any cases not can find one. If you have something in mind, please, share 🙏